src/ethereum/forks/amsterdam/state_tracker.py

```diff
@@ -467,37 +530,9 @@ def commit_transaction_frame(tx_frame: StateChanges) -> None:
         # else: Net-zero change within this transaction - skip
 
 
-def merge_on_failure(child_frame: StateChanges) -> None:
-    """
-    Merge child frame's changes into parent on failed completion.
-
-    Merges only read operations from the child frame into the parent.
-    Write operations are discarded since the frame reverted.
-    This is called when a call frame fails/reverts.
-
-    Parameters
-    ----------
-    child_frame :
-        The failed child frame.
-
-    """
-    assert child_frame.parent is not None
-    parent_frame = child_frame.parent
-    # Only merge reads and address accesses on failure
-    parent_frame.touched_addresses.update(child_frame.touched_addresses)
-    parent_frame.storage_reads.update(child_frame.storage_reads)
-
-    # Convert writes to reads (failed writes still accessed the slots)
-    for address, key, _idx in child_frame.storage_writes.keys():
-        parent_frame.storage_reads.add((address, key))
-
-    # Note: balance_changes, nonce_changes, and code_changes are NOT
-    # merged on failure - they are discarded
-
-
 def create_child_frame(parent: StateChanges) -> StateChanges:
     """
-    Create a child frame for nested execution.
+    Create a child frame linked to the given parent.
```

Comment:
The other thing we need to consider here is if we are ok with the `_block_access_index` defaulting to `0` for every child frame. I think every child frame should get the same `_block_access_index` as it parent.

Reply:
Agreed. I also think this can have the underscore removed? I will include this here.